### ISSUE TITLE
Vertical parallax

### ISSUE THREAD

**Issue by teaaaaaaaa** [chrismaltby/gb-studio#952]
**Is your feature request related to a problem? Please describe.**
Well the issue here is that if you attempt to scroll the screen vertically with parallax scrolling enabled, layers 1 and 2 will stay in the same vertical position and scrolling glitches will occur (even with only 1 layer).

**Describe the solution you'd like**
Combining the camera code with the interrupt code so that the layers can move vertically based off of the camera's Y position.

**Describe alternatives you've considered**
There are no alternatives.

**Additional context**

**Comment by darkhog** [comment]
Not easily possible due to the way Gameboy hardware does. You could do this with some vram tomfoolery, but it would require a major engine rewrite and wouldn't work that well.

**Comment by maxoakland** [comment]
> Not easily possible due to the way Gameboy hardware does. You could do this with some vram tomfoolery, but it would require a major engine rewrite and wouldn't work that well.

Would it be possible to “just” have the parallax layers scroll off the screen vertically, but not have any vertical parallax effect?

**Comment by teaaaaaaaa** [comment]
> Not easily possible due to the way Gameboy hardware does. You could do this with some vram tomfoolery, but it would require a major engine rewrite and wouldn't work that well.

By scrolling 8 pixel tall strips? Think you must be getting confused with the game gear and master system, which have this limitation and can only scroll 8 pixel tall strips. The gameboy, on the other hand, has complete control over where the background is scrolled on each scanline, and the final boss in wario land is an example of this.

**Comment by teaaaaaaaa** [comment]
> > Not easily possible due to the way Gameboy hardware does. You could do this with some vram tomfoolery, but it would require a major engine rewrite and wouldn't work that well.
> 
> Would it be possible to “just” have the parallax layers scroll off the screen vertically, but not have any vertical parallax effect? Enabling non-glitchy scenes with parallax that are taller than 144 and can scroll horizontally and vertically

I don't think so, because there's no such thing as negative scanlines (scanline #0 being the top of the screen) If the screen was scrolled vertically, the Y position of the 1st layer would probably underflow to 255 and possibly lead to a crash because of the Gameboy waiting for a scanline that doesn't exist. A simple range check would probably fix this though.

That's just my theory.

**Comment by Q-Bert-Reynolds** [comment]
[This Retro Game Mechanics Explained video](https://youtu.be/wt73KPS_23w) explains a few ways this could be achieved. The video focuses on NES, but the Game Boy works in a similar way.

If you go the background scrolling + scan line interrupt route, the full width of the screen will have the same scrolling. Within those full width horizontal chunks of screen, there's nothing stopping you from scrolling vertically though. This is how split screen multiplayer works. To really sell the vertical parallax though, you need to also constantly change the interrupt lines (ie. as you get higher on the screen, the horizon line should slowly get lower).

The tile replacement route is more difficult, but you could do pretty much anything you want with it. For a single tile, you could just shift the bits of your background tile. You could precompute the parallax tiles and replace them like you would any other animated tile. If you need "foreground" elements with bits of parallax peeking through, combine your parallax tile with the foreground tile using a mask.

**Comment by maxoakland** [comment]
I think it would be really cool to be able to do tile-based parallax
